message_to_fungible_token_predicate/src/main.sw

```diff
@@ -0,0 +1,119 @@
+predicate;
+
+use std::address::Address;
+use std::tx::*;
+use std::assert::assert;
+use std::hash::*;
+use std::contract_id::ContractId;
+
+const INPUT_COIN = 0u8;
+const INPUT_CONTRACT = 1u8;
+const INPUT_MESSAGE = 2u8;
+const OUTPUT_CONTRACT = 1u8;
+const OUTPUT_CHANGE = 3u8;
+const OUTPUT_VARIABLE = 4u8;
+
+// TO DO: use helper functions from std::tx when ready
+
+/// Read 256 bits from memory at an offset from a given pointer
+fn read_b256_from_pointer_offset(pointer: u32, offset: u32) -> b256 {
+    asm(buffer, ptr: pointer, off: offset) {
+        // Need to skip over `off` bytes
+        add ptr ptr off;
+        // Save old stack pointer
+        move buffer sp;
+        // Extend stack by 32 bytes
+        cfei i32;
+        // Copy 32 bytes
+        mcpi buffer ptr i32;
+        // `buffer` now points to the 32 bytes
+        buffer: b256
+    }
+}
+
+/// Get the type of an input at a given index
+fn get_input_type(index: u8) -> u8 {
+    let ptr = tx_input_pointer(index);
+    let input_type = tx_input_type(ptr);
+    input_type
+}
+
+/// Get the type of an output at a given index
+fn get_output_type(index: u8) -> u8 {
+    let ptr = tx_output_pointer(index);
+    let output_type = tx_output_type(ptr);
+    output_type
+}
+
+/// Get the script spending the input belonging to this predicate hash.
+/// Length is implicitly provided by casting to an array of u64. The bytecode is thus right-padded to the next full word
+fn get_script_bytecode<T>() -> T {
+    let script_ptr = tx_script_start_offset();
+    let script = asm(r1: script_ptr) {
+        r1: T
+    };
+    script
+}
+
+/// Get the ID of a contract input
+fn verify_and_get_contract_input_contract_id(index: u8) -> ContractId {
```

Comment:
In this particular case I'll side with Syd's suggested renaming, because getting a contract ID from an input isn't actually something that makes sense for non-contract inputs. In other words, you can't get the contract ID (safely) _without_ performing the check. Something like this would be more concise and allow for more thorough documentation:

```
/// Get the ID of a contract input.
///
/// # Reverts
///
/// - If input `index` is not a contract input.
```

Reply:
That's fine. The context for my overly verbose naming was that previously there was a function that did this _unsafely_ (the comments actually warned about undefined behaviour in case of a non-contract input).

But since this replaces that, we can give it a sensible name. We want users to be able to expect safe behaviour by default. I'll make the change and also make sure that this is uniform over the library 👍